lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -427,6 +437,32 @@ newtype ApiMaintenanceAction = ApiMaintenanceAction
     { gcStakePools :: ApiT PoolMetadataGCStatus
     } deriving (Eq, Generic, Show)
 
+data ApiAsset = ApiAsset
+    { policyId :: ApiT W.TokenPolicyId
+    , assetName :: ApiT W.TokenName
+    , displayName :: Text
+    , metadata :: Maybe (ApiT W.AssetMetadata)
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+toApiAsset :: Maybe W.AssetMetadata -> W.AssetId -> ApiAsset
+toApiAsset metadata_ (W.AssetId policyId_ assetName_) = ApiAsset
+    { policyId = ApiT policyId_
+    , assetName = ApiT assetName_
+    , displayName = makeDisplayName policyId_ assetName_ metadata_
+    , metadata = ApiT <$> metadata_
+    }
+
+makeDisplayName :: W.TokenPolicyId -> W.TokenName -> Maybe W.AssetMetadata -> Text
+makeDisplayName pid an = \case
+    Just (W.AssetMetadata md) -> md
+    Nothing -> case T.decodeUtf8' assetNameBytes of
+        Right "" -> hexText $ getHash $ W.unTokenPolicyId pid
+        Right text -> text
+        Left _ -> hexText assetNameBytes
```

Comment:
I am not sure to get the rationale for the `Nothing` branch in this function. The `assetName` isn't something that is supposed to be user-facing or displayed to user. 

I'd suggest to leave it `null` if there's no metadata (that is, simply leave it as `metadata.name`) or, to go with your first idea in the cryptpad spec: a bech32 encoded string comprised of the policyId | assetName, possibly hashed. Though, it's not really anymore displayable than the policyId and assetName.

Reply:
Ok I have removed displayName for the time being.